aten/src/ATen/native/cpu/SumKernel.cpp

```diff
@@ -0,0 +1,304 @@
+    int64_t size0, int64_t size1) {
+  using vec_t = Vec256<scalar_t>;
+  constexpr int64_t nrows = 4;
+  constexpr int64_t vec_stride = vec_t::size() * sizeof(scalar_t);
+
+  // Input is contiguous over the second (non-reduced) dimension
+  int64_t j = 0;
+  for (; j + nrows * vec_t::size() <= size1; j += nrows * vec_t::size()) {
+    const auto *row_in = data[1] + j * sizeof(scalar_t);
+    auto sums = multi_row_sum<vec_t, nrows>(row_in, inner_stride, vec_stride, size0);
+
+    for (int64_t i = 0; i < nrows; ++i) {
+      const int64_t base_idx = j + i * vec_t::size();
+
+      std::array<scalar_t, vec_t::size()> ans;
+      sums[i].store(ans.data());
+      store_result(data[0], out_stride, base_idx, ans);
+    }
+  }
+
+  for (; j + vec_t::size() <= size1; j += vec_t::size()) {
+    const auto *row_in = data[1] + j * sizeof(scalar_t);
+    const vec_t sums = row_sum<vec_t>(row_in, inner_stride, size0);
+
+    std::array<scalar_t, vec_t::size()> ans;
+    sums.store(ans.data());
+    store_result(data[0], out_stride, j, ans);
+  }
+
+  for (; j < size1; ++j) {
+    const auto *row_in = data[1] + j * sizeof(scalar_t);
+    scalar_t ans = row_sum<scalar_t>(row_in, inner_stride, size0);
+    store_result(data[0], out_stride, j, ans);
+  }
+}
+
+template <typename scalar_t>
+void scalar_outer_sum(
+    char * C10_RESTRICT data[2], int64_t in_strides[2], int64_t out_stride,
+    int64_t size0, int64_t size1) {
+
+  constexpr int64_t nrows = 4;
+  int64_t j = 0;
+  for (; j + (nrows - 1) < size1; j += nrows) {
+    const auto *row_in = data[1] + j * in_strides[1];
+    auto sums = multi_row_sum<scalar_t, nrows>(
+        row_in, in_strides[0], in_strides[1], size0);
+    store_result(data[0], out_stride, j, sums);
+  }
+
+  for (; j < size1; ++j) {
+    const auto *row_in = data[1] + j * in_strides[1];
+    scalar_t ans = row_sum<scalar_t>(row_in, in_strides[0], size0);
+    store_result(data[0], out_stride, j, ans);
+  }
+}
+
+void sum_kernel_impl(TensorIterator &iter) {
+  if (isIntegralType(iter.dtype(), /*includeBool=*/ true)) {
+    AT_DISPATCH_INTEGRAL_TYPES_AND(ScalarType::Bool, iter.dtype(), "sum_cpu",
+      [&] {
+        binary_kernel_reduce_vec(
+            iter, [=](scalar_t a, scalar_t b) -> scalar_t { return a + b; },
+            [=](Vec256<scalar_t> a, Vec256<scalar_t> b) { return a + b; });
+      });
+    return;
+  }
+
+  // Custom floating point sum for better accuracy
+  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND2(
+    ScalarType::BFloat16, ScalarType::Half, iter.dtype(), "sum_cpu",
+    [&] {
+      iter.output().fill_(scalar_t(0));
+      iter.parallel_reduce(
+        [&](char** data, const int64_t* strides, int64_t size0, int64_t size1) {
+          int64_t in_strides[] = { strides[1], strides[3] };
+          int64_t out_strides[] = { strides[0], strides[2] };
+
+          // Move reduction to be the 1st dim
+          if (out_strides[0] != 0 && out_strides[1] == 0) {
```

Comment:
This doesn't seem to always happen. For example, 
```python
torch.rand(100, 1).sum(1)
```
Gives me strides `4, 0`.

Reply:
hmhm, that's probably because when you are reducing over a dimension of size 1, is_dim_reduced does not identify it as reduction dimension.
https://github.com/pytorch/pytorch/blob/4c7d81f8479bce320cc11d1eb3adaf8ab0b90099/aten/src/ATen/native/TensorIterator.cpp#L472-L479
TensorIterator logic around reductions is pretty confusing :-(

Reply:
For the `basic_loop` version of the reduction, that would give a better memory access pattern over the input vector. I'm guess that's why it's done this way?

Reply:
I've dug a bit deeper, and it turns out in this case dimensions are coalesced in TI, so instead of (100,1) tensor you have (100) tensor. Then bogus (0,0) strides are pushed back to the end of the strides vector to fill in, and a bogus size1=1 is sent to the loop. Depending on whether input tensor is contiguous or not, vectorized_outer_sum or scalar_inner_sum will be called, both will produce correct results. Still don't know how out_strides can both be non-zero, but maybe it happens when `out` kwarg is specified and for some reason strides for dimension equal to 1 are not reset (their values don't matter).

Reply:
I found a minimal reproducer for the non-zero strided reduction. It also relies on reduction over a length-1 dimension:
```python
torch.rand(2).expand(1, 3, 2).sum(0)
```